### ISSUE TITLE
IOException running NuGet-Migrations during tests in dotnet CLI first run

### ISSUE THREAD

**Issue by akoeplinger** [dotnet/runtime#80619]
This is affecting 6.0, 7.0 and 8.0.

In https://github.com/dotnet/runtime/pull/80510#issuecomment-1382160934 we saw a failure because some NuGet migration code failed to run during the dotnet CLI first run experience:

```
+ dotnet /datadisks/disk1/work/ADFC09B1/p/xunit/xunit.console.dll JIT/HardwareIntrinsics/JIT.HardwareIntrinsics.XUnitWrapper.dll -parallel collections -nocolor -noshadow -xml testResults.xml -trait TestGroup=JIT.HardwareIntrinsics.Arm.ArmBase
Microsoft.DotNet.XUnitConsoleRunner v2.5.0 (64-bit .NET 7.0.2)
  Discovering: JIT.HardwareIntrinsics.XUnitWrapper (method display = ClassAndMethod, method display options = None)
  Discovered:  JIT.HardwareIntrinsics.XUnitWrapper (found 6 of 362 test cases)
  Starting:    JIT.HardwareIntrinsics.XUnitWrapper (parallel test collections = on, max threads = 2)
    JIT/HardwareIntrinsics/Arm/ArmBase.Arm64/ArmBase.Arm64_ro/ArmBase.Arm64_ro.sh [FAIL]
      System.IO.IOException: The system cannot open the device or file specified. : 'NuGet-Migrations'
         at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew)
         at System.Threading.Mutex..ctor(Boolean initiallyOwned, String name)
         at NuGet.Common.Migrations.MigrationRunner.Run()
         at Microsoft.DotNet.Configurer.DotnetFirstTimeUseConfigurer.Configure()
         at Microsoft.DotNet.Cli.Program.ConfigureDotNetForFirstTimeUse(IFirstTimeUseNoticeSentinel firstTimeUseNoticeSentinel, IAspNetCertificateSentinel aspNetCertificateSentinel, IFileSentinel toolPathSentinel, Boolean isDotnetBeingInvokedFromNativeInstaller, DotnetFirstRunConfiguration dotnetFirstRunConfiguration, IEnvironmentProvider environmentProvider, Dictionary`2 performanceMeasurements)
         at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, TimeSpan startupTime, ITelemetry telemetryClient)
         at Microsoft.DotNet.Cli.Program.Main(String[] args)
      
      Return code:      1
```

We should look into disabling the first run experience via the `DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1` env var, it doesn't make sense to run that for the tests.

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[368683](https://dev.azure.com/dnceng-public/public/_build/results?buildId=368683)|dotnet/runtime|[JIT.Methodical.eE.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=368683&view=ms.vss-test-web.build-test-results-tab&runId=7815400&resultId=102496)|dotnet/runtime#90241|
|[362596](https://dev.azure.com/dnceng-public/public/_build/results?buildId=362596)|dotnet/runtime|[JIT.Methodical.eE.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=362596&view=ms.vss-test-web.build-test-results-tab&runId=7661116&resultId=102496)|dotnet/runtime#89003|
|[361655](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361655)|dotnet/runtime|[JIT.Math.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361655&view=ms.vss-test-web.build-test-results-tab&runId=7638022&resultId=102855)|dotnet/runtime#89905|
|[361259](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361259)|dotnet/runtime|[JIT.Generics.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361259&view=ms.vss-test-web.build-test-results-tab&runId=7627032&resultId=102496)||
|[361075](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361075)|dotnet/runtime|[JIT.Regression.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361075&view=ms.vss-test-web.build-test-results-tab&runId=7622388&resultId=102856)|dotnet/runtime#89867|
|[360768](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360768)|dotnet/runtime|[JIT.Regression.JitBlue.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360768&view=ms.vss-test-web.build-test-results-tab&runId=7614654&resultId=102855)|dotnet/runtime#89815|
|[360148](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360148)|dotnet/runtime|[JIT.jit64.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360148&view=ms.vss-test-web.build-test-results-tab&runId=7598234&resultId=102496)|dotnet/runtime#89814|
|[359522](https://dev.azure.com/dnceng-public/public/_build/results?buildId=359522)|dotnet/runtime|[JIT.Methodical.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=359522&view=ms.vss-test-web.build-test-results-tab&runId=7580578&resultId=102496)|dotnet/runtime#89805|
|[356421](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356421)|dotnet/runtime|[JIT.Methodical.a-dA-D.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356421&view=ms.vss-test-web.build-test-results-tab&runId=7498350&resultId=102496)||
|[355235](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355235)|dotnet/runtime|[JIT.Methodical.f-iF-I.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355235&view=ms.vss-test-web.build-test-results-tab&runId=7471180&resultId=102496)|dotnet/runtime#89009|
|[343534](https://dev.azure.com/dnceng-public/public/_build/results?buildId=343534)|dotnet/runtime|[JIT.Methodical.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=343534&view=ms.vss-test-web.build-test-results-tab&runId=7177384&resultId=102496)||
|[342037](https://dev.azure.com/dnceng-public/public/_build/results?buildId=342037)|dotnet/runtime|[JIT.Regression.JitBlue.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=342037&view=ms.vss-test-web.build-test-results-tab&runId=7142106&resultId=102855)|dotnet/runtime#89008|
|[341937](https://dev.azure.com/dnceng-public/public/_build/results?buildId=341937)|dotnet/runtime|[JIT.Methodical.a-dA-D.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=341937&view=ms.vss-test-web.build-test-results-tab&runId=7140062&resultId=102496)|dotnet/runtime#89003|
|[337982](https://dev.azure.com/dnceng-public/public/_build/results?buildId=337982)|dotnet/runtime|[JIT.HardwareIntrinsics.X86.Avx2.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=337982&view=ms.vss-test-web.build-test-results-tab&runId=7032408&resultId=102855)|dotnet/runtime#88809|
|[337742](https://dev.azure.com/dnceng-public/public/_build/results?buildId=337742)|dotnet/runtime|[JIT.Regression.CLR-x86-JIT.V1-M12-M13.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=337742&view=ms.vss-test-web.build-test-results-tab&runId=7025630&resultId=102496)|dotnet/runtime#88769|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|1|15|

### Known Issue Error Message
Fill the error message using [known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorPattern": "The system cannot open the device or file specified. : (&#39;|')NuGet-Migrations(&#39;|')",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** 
**Result validation: :warning:** Validation could not be done without an Azure DevOps build URL on the issue. Please add it to the "**Build: :mag_right:**" line.
**Validation performed at:** 6/28/2023 10:04:55 PM UTC

**Comment by ViktorHofer** [comment]
I thought that we were already setting that property at least for libraries tests. Maybe not for runtime tests? cc @jkoritzinsky

**Comment by carlossanlop** [comment]
Another PR that hit this: https://github.com/dotnet/runtime/pull/80615

**Comment by radical** [comment]
Another [rolling build](https://dev.azure.com/dnceng-public/public/_build/results?buildId=160649)([log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-38d0b7e852fe454f88/JIT.Regression.JitBlue/1/console.c8c670d4.log?helixlogtype=result)).

**Comment by carlossanlop** [comment]
Another 7.0 dependency flow PR affected by this: https://github.com/dotnet/runtime/pull/81812

**Comment by carlossanlop** [comment]
> I thought that we were already setting that property at least for libraries tests. Maybe not for runtime tests? cc @jkoritzinsky

Ping @jkoritzinsky

**Comment by lewing** [comment]
did we do something in net8 that needs to also be done in net7 here?

**Comment by radical** [comment]
This is affecting lot of builds.

**Comment by trylek** [comment]
I don't see any property / variable named <code>DOTNET_SKIP_FIRST_TIME_EXPERIENCE</code> in the dotnet/sdk repo, @ViktorHofer / @akoeplinger - am I looking in the wrong place (can you please point me to the place where it's used?) or are you saying we need to build this as a new feature? In fact I'm somewhat puzzled by the above code, these are apparently CoreCLR tests and they shouldn't be normally executed via dotnet, that must be coming from somewhere else like Crossgen2 execution.

**Comment by radical** [comment]
https://github.com/dotnet/sdk/blob/main/documentation/manpages/sdk/dotnet-environment-variables.7#L296 Was it removed?

**Comment by trylek** [comment]
This one is interesting but it's not the root thing, it just reacts on the fact that the variable isn't used anywhere:

https://github.com/dotnet/sdk/commit/27b6661ce31434055287d802929c5ba61c82ddb4

I continue looking what gutted this behavior.

**Comment by trylek** [comment]
OK, so as I now understand it, the call to <code>MigrationRunner.Run</code> that is causing the above issue is now here:

https://github.com/dotnet/sdk/blob/fbc089ea96f0187c31e9ae176278334c8d5defc6/src/Cli/Microsoft.DotNet.Configurer/DotnetFirstTimeUseConfigurer.cs#L62

The check <code>ShouldPrintFirstTimeUseNotice()</code> basically calls the method <code>Exists</code> on the <code>IFirstTimeUseNoticeSentinel</code> interface. It seems to me that it should be possible to pull this off by using the

https://github.com/dotnet/sdk/blob/main/src/Cli/Microsoft.DotNet.Configurer/NoOpFirstTimeUseNoticeSentinel.cs

class for the implementation of the sentinel; this seems to be tied to the conditional block

https://github.com/dotnet/sdk/blob/b68de63954610d17b82166de586567dd244a85ca/src/Cli/dotnet/Program.cs#L183

triggered by the command-line parameter

https://github.com/dotnet/sdk/blob/12209f087e1c0256db8b08e7e7217f7526628d8b/src/Cli/dotnet/Parser.cs#L32

which leads me to

https://github.com/dotnet/sdk/blob/2c98ff1381def385256788cfcbaefb3ac32f5778/src/Cli/dotnet/commands/dotnet-internal-reportinstallsuccess/InternalReportinstallsuccessCommandParser.cs#L10

In this light I tend to believe that it might be possible to workaround this by using the internal option <code>internal-reportinstallsuccess</code>. I must admit I'm seeing this logic for the first time so it would be great to hear confirmation from someone familiar with this logic, actually fixing this e.g. w.r.t. Crossgen2 execution in CoreCLR tests is trivial with this background.

**Comment by lewing** [comment]
quite a few things expect it to work https://github.com/search?q=org%3Adotnet+DOTNET_SKIP_FIRST_TIME_EXPERIENCE&type=code

**Comment by trylek** [comment]
As I said, I'm not all that much familiar with this era of SDK development. To reinstate the variable, I guess it would be easiest to hack around

https://github.com/dotnet/sdk/blob/b68de63954610d17b82166de586567dd244a85ca/src/Cli/dotnet/Program.cs#L168

where we're already inspecting several environment variables. Adding an extra variable representing the "no-first-time-experience" and using it in an OR clause at line

https://github.com/dotnet/sdk/blob/b68de63954610d17b82166de586567dd244a85ca/src/Cli/dotnet/Program.cs#L180

would probably go a long way towards satisfying both behaviors. Having said that, I still maintain I'm not an SDK expert so I'd love to hear from at least one person familiar with the repo that my conclusions make some sense.

**Comment by trylek** [comment]
@marcpopMSFT - could you please comment on this or route this to someone familiar with the code in question?

**Comment by radekdoulik** [comment]
Looks similar to https://github.com/NuGet/Home/issues/12159

**Comment by akoeplinger** [comment]
Hmm interesting, looks like it was removed indeed: https://github.com/dotnet/sdk/issues/9945.
Given it is referenced a lot in different places and we have the nuget migration thing during first run I'd say it might be good to bring it back.

**Comment by pavelsavara** [comment]
Another exhibit on https://github.com/dotnet/runtime/pull/82834  [Log](https://helix.dot.net/api/2019-06-17/jobs/8bbff426-c770-40ac-aa56-2e26c21f1600/workitems/NoWorkload-Wasm.Build.Tests/console)

**Comment by lewing** [comment]
@marcpopMSFT this is breaking a lot of builds can we get status on it?

**Comment by marcpopMSFT** [comment]
The call to NuGet.Common.Migrations.MigrationRunner.Run was added at the request of the nuget team. CC @aortiz-msft @kartheekp-ms to investigate why it's failing.

We could potentially add a variable on whether to run this check but it was added for security reasons so that would be there just for scenarios like this. Is it possible to create the sentinel file before running the tests?

**Comment by steveisok** [comment]
@trylek @akoeplinger do you know if a sentinel file is possible?

**Comment by trylek** [comment]
I think I can probably imagine some logic that would trigger this before running the actual tests, perhaps as a prolog of the generated legacy XUnit wrappers (for non-merged tests) and the new-style Roslyn-generated full wrappers for the merged tests but it's unclear to me whether it would fix the problem or just make it less frequent. If it's the latter I think we should rather try to get to the bottom of why this step is failing on the Helix machines and / or try to find consensus regarding its suppression for the Helix tests. It also doesn't answer the concern raised by @lewing, namely that removal of the <code>DOTNET_SKIP_FIRST_TIME_EXPERIENCE</code> flag apparently wasn't fully synchronized with other repos that still set it in their scripts so we still need to figure out whether we can bring it back or whether we need to audit all places that are using it and somehow bring them in sync with its removal; /cc @jkoritzinsky for additional thoughts.

**Comment by trylek** [comment]
That sounds good to me and shouldn't be a hard change to make assuming we manage to convince ourselves that it indeed fixes the problem, not just moves it to a different place.

**Comment by marcpopMSFT** [comment]
I should note that nuget will run that code the first time a restore command is issued so any tests that do a restore will also hit that code path. There does need to be a separate investigation as to why that code path is failing as it was for migrating machines from an older, unsecure configuration.

Note that that flag shows up a few hundred times in the dotnet org: https://github.com/search?p=1&q=org%3Adotnet+DOTNET_SKIP_FIRST_TIME_EXPERIENCE&type=Code  CC @baronfel

**Comment by trylek** [comment]
I believe this slightly differs between coreclr / mono and library tests - library tests use "dotnet vstest" so they run dotnet to restore and build the actual tests. CoreCLR and Mono runtime testing currently doesn't use "dotnet vstest", it uses a custom test execution infrastructure, in general the tests get built and published on the AzDO build machines, AFAIK in Helix dotnet is only used to execute the managed tools - Crossgen2, R2RDump and probably R2RTest which is used in a handful of readytorun tests.

**Comment by radical** [comment]
> I believe this slightly differs between coreclr / mono and library tests - library tests use "dotnet vstest" so they run dotnet to restore and build the actual tests. CoreCLR and Mono runtime testing currently doesn't use "dotnet vstest", it uses a custom test execution infrastructure, in general the tests get built and published on the AzDO build machines, AFAIK in Helix dotnet is only used to execute the managed tools - Crossgen2, R2RDump and probably R2RTest which is used in a handful of readytorun tests.

Except for Wasm.Build.Tests which does full project builds, including restores.

**Comment by kartheekp-ms** [comment]
A similar issue was reported in the past https://github.com/dotnet/runtime/pull/78062#issuecomment-1314259386.  May be this error has the same root cause mentioned in https://github.com/dotnet/runtime/issues/36823 which was created as a follow-up for https://github.com/PowerShell/PSReadLine/issues/1464

cc @marcpopMSFT @aortiz-msft

**Comment by lewing** [comment]
For additional information the wasm lanes run the runtime/jit tests using a wrapper shell and they can run concurrently so they tend to trigger race conditions in the build system.  The part that is failing isn't related to the wasm runtime itself.

**Comment by radical** [comment]
FYI, publishing wasm projects does a nested build - using the `MSBuild` task. So it applies to almost all the wasm test builds that we have.

**Comment by lewing** [comment]
the actual failure looks a lot like https://github.com/dotnet/runtime/issues/76736

**Comment by lewing** [comment]
cc @kouvel

**Comment by lewing** [comment]
@stephentoub any idea what might cause mutex to fail like this?

**Comment by stephentoub** [comment]
Mutex on Linux needs to perform a variety of I/O operations, including creating a file on disk and creating a pthreads mutex; it's possible one of those is failing for some reason.

That error string looks like https://github.com/dotnet/runtime/blob/622277e02da5517aaeaa322be27079822b4d33e2/src/coreclr/pal/src/misc/errorstrings.cpp#L71, which is used in a variety of places.  For example, it'll be used in https://github.com/dotnet/runtime/blob/622277e02da5517aaeaa322be27079822b4d33e2/src/coreclr/pal/src/synchobj/mutex.cpp#L791 if the operation encounters an EPERM errno from pthread_mutex_init.

@kouvel?

**Comment by janvorli** [comment]
However, it is possible that the current issue is caused by some other reason. If the issue is easily reproducible in the ci, it would make sense to create a testing PR with added listing of the directory tree and their permissions / users under the /tmp/.dotnet/shm and dumping the current session id. That would reveal if the issue is what I've mentioned or something different.

**Comment by steveisok** [comment]
@trylek @jkoritzinsky @akoeplinger can one of you try out Jan's suggestion? 

Should this also be an issue in NuGet?

**Comment by kouvel** [comment]
It would probably be useful to generally preserve more information about the error, at least the last call, some parameters, and the error code. I'll look into doing that.

If above turns out to be the issue, a couple of options may be:
- After a container is initialized/updated and before it will be reused, as a last step in the Dockerfiles perhaps the /tmp/.dotnet/shm could be deleted
- Using a global mutex with "Global\" prefix in NuGet may bypass the issue, if that would be appropriate, as global mutexes are intended to be used by multiple users/sessions

**Comment by kartheekp-ms** [comment]
> Using a global mutex with Global\ prefix in NuGet may bypass the issue, if that would be appropriate, as global mutexes are intended to be used by multiple users/sessions

The end result of this migration process is to fix folders permissions respecting users `umask` and create a file under `home(userprofile)/.local/share/NuGet/Migrations` so that next NuGet operation don't run the migration logic again. Given that this migration is specific to user we chose a `local mutex`.

Following the example from https://learn.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=net-6.0#remarks. I am wondering if we should create local mutex and assign it to a static variable instead of creating it whenever `Run()` method is invoked and also add a destructor to dispose the Mutex. 

https://github.com/NuGet/NuGet.Client/blob/3ab7d1462029e0b211a88040c952c9b8a679708d/src/NuGet.Core/NuGet.Common/Migrations/MigrationRunner.cs#LL23C17-L23C73

```cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

using System;
using System.IO;
using System.Threading;

namespace NuGet.Common.Migrations
{
    public static class MigrationRunner
    {
        private const string MaxMigrationFilename = "1";
        private static Mutex mutex = new Mutex(false, "NuGet-Migrations");

        public static void Run()
        {
            string migrationsDirectory = GetMigrationsDirectory();
            var expectedMigrationFilename = Path.Combine(migrationsDirectory, MaxMigrationFilename);

            if (!File.Exists(expectedMigrationFilename))
            {
                  if (WaitForMutex(mutex))
                    {
                        try
                        {
                            // Existing code
                        }
                        catch {}
                        finally
                        {
                            mutex.ReleaseMutex();
                        }
                    }
            }

            static bool WaitForMutex(Mutex mutex)
            {
                bool captured;

                try
                {
                    captured = mutex.WaitOne(TimeSpan.FromMinutes(1), false);
                }
                catch (AbandonedMutexException)
                {
                    captured = true;
                }

                return captured;
            }
        }
      
      ~MigrationRunner()
          {
             mutex.Dispose();
          }
    }
}
```

Does this work around unblock the current issue until we have a proper fix in the runtime repo?

**Comment by kouvel** [comment]
> The end result of this migration process is to fix folders permissions respecting users `umask` and create a file under `home(userprofile)/.local/share/NuGet/Migrations` so that next NuGet operation don't run the migration logic again. Given that this migration is specific to user we chose a `local mutex`. My understanding is that if we choose global mutex multiple users can't run NuGet migrations logic concurrently.

A local mutex is session-local and not necessarily tied to a user. For instance on Linux, a user can have multiple terminal shells open and each of those would have a different session ID, so a session-local mutex would not protect code running as the same user concurrently under different shells. On Windows, it's possible to run processes as a particular user under different sessions, for instance a service and a process in a login session can run as the same user. Perhaps a global mutex would work, or maybe using file operations in a user-specific folder to mimic a user-specific lock.

**Comment by kouvel** [comment]
> `~MigrationRunner()`

I don't think a static class can have a finalizer since it can't be instantiated. I don't think there would be any significant difference between using a static field or creating a mutex on the fly. The timings would change but something is going wrong that I don't understand yet.

**Comment by kouvel** [comment]
Does anyone know if there's a way to get access to the container image used to run a particular test in the CI? The test is failing fairly reliably, and having access to that container image would likely enable me to repro this more easily and do some logging.

**Comment by hoyosjs** [comment]
@kouvel if you need help setting up an environment, feel free to ping me. There's a test https://dev.azure.com/dnceng-public/public/_build/results?buildId=200752&view=ms.vss-test-web.build-test-results-tab&runId=3764716&resultId=103713&paneView=debug

> "HelixJobId": "08f6d861-acc6-4905-be8a-e8a8e2e6c62a",
"HelixWorkItemName": "JIT.Math"

That fails directly on the VM. It's easy enough to set up a VM with the same image and number of cores in the helix-repro lab. The image is ubuntu.1804.amd64 and it's a 2 core D_v4 (D2a_v4 I believe).

**Comment by mxschmitt** [comment]
For folks who run into it, I was able to workaround it by:

```
rm -rf /tmp/.dotnet/
```

(or give your user permissions to this folder).

**Comment by akoeplinger** [comment]
> Does anyone know if there's a way to get access to the container image used to run a particular test in the CI?

If you look at the first line of the console log it will say something like this if the test was running in Docker:

```
Console log: 'System.Runtime.Serialization.Xml.Tests' from job 92398c09-eb7f-4de3-89b5-e52616c28233 (ubuntu.2004.amd64.open.rt) using docker image mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04-helix-wasm-amd64 on a0000M2
running $HELIX_CORRELATION_PAYLOAD/scripts/3db7317da8c649a8ac1b933ed50769e0/execute.sh in /datadisks/disk1/work/A9A208FA/w/B1F30963/e max 1800 seconds
```

In that case you'd need to use the `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04-helix-wasm-amd64` image.

**Comment by tte-schleupen** [comment]
Is there a fix for this problem? Im having the same problem inside a docker container. Tried the workarounds, especially the "DOTNET_SKIP_FIRST_TIME_EXPERIENCE = 1", but nothing worked. Any ideas? Is there a way to skip the migration?

**Comment by mxschmitt** [comment]
@tte-schleupen https://github.com/dotnet/runtime/issues/80619#issuecomment-1465768181 should work for you.

**Comment by kouvel** [comment]
Thanks @akoeplinger! I looked at the base container image and didn't find any issues with it, the /tmp/.dotnet directory doesn't exist. There appears to be something that is run on the base container image between its initial state and to the point where the console log begins, I wasn't able to find that info from the data I downloaded from runfo for the job / work item, and wasn't able to repeat the test run or repro the issue in the same way as in the CI. It looks like it is using a .NET 8 daily SDK, so perhaps a change could be made and merged to retain more info about the system call failure to shed more light on future runs/failures.

**Comment by mxschmitt** [comment]
Minimal reproducible bisected to good 6.0.402 bad 6.0.403:

```Dockerfile
# Bad:
FROM mcr.microsoft.com/dotnet/sdk:6.0.402-alpine3.16
# Good: FROM mcr.microsoft.com/dotnet/sdk:6.0.403-alpine3.16

RUN adduser --disabled-password --gecos '' pwuser

RUN mkdir /tmp/demo-app && \
    cd /tmp/demo-app && \
    dotnet new console && \
    dotnet add package MSTest.TestAdapter && \
    dotnet build

USER pwuser

RUN mkdir /home/pwuser/app && cd /home/pwuser/app && dotnet new console

```

**Comment by jcolebrand** [comment]
> However, it is possible that the current issue is caused by some other reason. If the issue is easily reproducible in the ci, it would make sense to create a testing PR with added listing of the directory tree and their permissions / users under the /tmp/.dotnet/shm and dumping the current session id. That would reveal if the issue is what I've mentioned or something different.

This is 100% the issue, I've been getting this a few times and can reproduce exactly the issue. If I first run pwsh as root (simply reproducible scenario) and force quit that shell so it doesn't neatly close things out, and log in as another user and run pwsh that doesn't have any privileges, the /tmp/.dotnet is locked without being writable.

This is runtime related, and I think it's this specific scenario being encountered: https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/sharedmemory/sharedmemory.cpp#L127-L132

**Comment by lewing** [comment]
This is still breaking a large number of wasm builds and other failures are starting to slip in because of the merging on red.  We need to resolve the regression.

**Comment by steveisok** [comment]
> Why is this process ran on a newly created docker container ? Is the beginning state of the container not suitable for nuget ? Also, wouldn't be possible to make this process work without locking ?

@akoeplinger is there something we can do with the wasm container to try and work around this?

**Comment by steveisok** [comment]
> @steveisok we could probably precreate `/tmp/.dotnet/` in the wasm container and give it some broad permissions...

Ok, can you try that out to see if it can unblock @lewing?

**Comment by janvorli** [comment]
I would like to clarify few things. This discussion seems to be mixing two issues:
1. The one we see in the CI where the reason seems to be unknown based on @kouvel's comments
2. A known and expected issue that stems from the permission fix (mentioned above) on the /tmp/.dotnet/shm directory that happens when:
    * a privileged user runs dotnet app that uses named mutexes with local name (not prefixed or prefixed by `Local\`), so dotnet creates a /tmp/.dotnet/session**XYZ** directory where **XYZ** is a session id
    * the mutex outlives the creating app, either due to a bug in the app, the app getting killed, etc.
    * then a non-privileged user runs dotnet app and by a chance gets the same session id, which is usually the process id of the login shell
    * that app also creates a named mutex with local name, so dotnet tries to create the /tmp/.dotnet/session**XYZ** directory where again, the **XYZ** is a session id
    * since that directory already exists and belongs to a privileged user, the current user's app fails to access it and fails to create the named mutex

The second issue occurs rarely when running outside of a docker as the session id clashes are rare. However, when running in docker, the session id is always zero (or 1, I don't remember it by heart). So e.g. when the docker image is built with the user set to root (the default) and during the built it runs a dotnet app that creates and doesn't close a named mutex with a local name and the created image is set to run code as non-root user (which has to be explicitly set), we get the problem.

While we could easily solve the second issue by including user id in the session directory name, it turns out that it would break ability of apps running with older .NET runtime to share named mutex with local name with apps running with a newer .NET runtime that would have such a change.

**Comment by kouvel** [comment]
> So we still need to find out case 1 then. @kouvel do you have a test PR somewhere that I could take a look, maybe I can figure out what exactly Helix runs.

My test PR is: https://github.com/dotnet/runtime/pull/83186, and the failed run I was looking at is https://github.com/dotnet/runtime/runs/11994338068.

> we could probably precreate `/tmp/.dotnet/` in the wasm container and give it some broad permissions...

The issue could also occur due to the permissions of the session directory under `/tmp/.dotnet/shm/`, so not sure that would work, though I don't know for sure what the issue is that is occurring in the CI.

**Comment by lewing** [comment]
@akoeplinger @steveisok  Workarounds did not resolve the issue.

**Comment by akoeplinger** [comment]
@lewing do you have an example build?

**Comment by lewing** [comment]
> @lewing do you have an example build?

no @steveisok pointed out it appears to be just the runtime tests failing now

**Comment by am11** [comment]
> We should look into disabling the first run experience via the `DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1` env var, it doesn't make sense to run that for the tests.

fyi: `DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1` lookup was removed in https://github.com/dotnet/cli/commit/7efbf4770630fd370779d222ce9aba35ffe56100 and brought back as `DOTNET_NOLOGO=1` in https://github.com/dotnet/sdk/commit/a41c9135e5e23440e106930c7ce949ded2ab3d26. Former is unsupported in all active releases of SDK.

```sh
$ find ~/.dotnet -name '*dotnetFirstUseSentinel' -exec rm {} \;
$ dotnet --help
# shows telemetry banner

$ find ~/.dotnet -name '*dotnetFirstUseSentinel' -exec rm {} \;
$ DOTNET_NOLOGO=1 dotnet --help
# no banner
```

**Comment by akoeplinger** [comment]
@am11 yep unfortunately the nuget migrations code runs before the NoLogo check so we can't use that one: https://github.com/dotnet/sdk/blob/ec824c9797659318c772330cb176286643a25dfe/src/Cli/Microsoft.DotNet.Configurer/DotnetFirstTimeUseConfigurer.cs#L62-L64

**Comment by ericstj** [comment]
Hit in https://github.com/dotnet/runtime/pull/84355

**Comment by lewing** [comment]
hit in https://github.com/dotnet/runtime/pull/84420 but pattern didn't match

**Comment by lewing** [comment]
Is there a plan for fixing the underlying issue here, a mutex that deadlocks for compatibility reasons still is a very broken mutex.

**Comment by kouvel** [comment]
The current plan after discussion is to revert the previous changes that made session-local mutexes user-specific, and offer user-specific mutexes as a new feature in a future release.
